src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsWriter.java

```diff
@@ -0,0 +1,216 @@
+     * @param componentName the component name for the SortedTermsMeta
+     * @param metadataWriter the MetadataWriter for storing the SortedTermsMeta
+     * @param termsData where to write the prefix-compressed terms data
+     * @param termsDataBlockOffsets  where to write the offsets of each block of terms data
+     * @param trieWriter where to write the trie that maps the terms to point ids
+     */
+    public SortedTermsWriter(String componentName,
+                             MetadataWriter metadataWriter,
+                             IndexOutput termsData,
+                             NumericValuesWriter termsDataBlockOffsets,
+                             IndexOutputWriter trieWriter) throws IOException
+    {
+        this.componentName = componentName;
+        this.metadataWriter = metadataWriter;
+        this.trieOutput = trieWriter;
+        SAICodecUtils.writeHeader(this.trieOutput);
+        this.trieWriter = new IncrementalDeepTrieWriterPageAware<>(trieSerializer, trieWriter.asSequentialWriter());
+        SAICodecUtils.writeHeader(termsData);
+        this.termsOutput = termsData;
+        this.bytesStartFP = termsData.getFilePointer();
+        this.offsetsWriter = termsDataBlockOffsets;
+    }
+
+    /**
+     * Appends a term at the end of the sequence.
+     * Terms must be added in lexicographic order.
+     *
+     * @throws IOException if write to disk fails
+     * @throws IllegalArgumentException if the term is not greater than the previous added term
+     */
+    public void add(final @Nonnull ByteComparable term) throws IOException
+    {
+        tempTerm.clear();
+        copyBytes(term, tempTerm);
+
+        final BytesRef termRef = tempTerm.get();
+        final BytesRef prevTermRef = this.prevTerm.get();
+
+        Preconditions.checkArgument(prevTermRef.length == 0 || prevTermRef.compareTo(termRef) < 0,
+                                    "Terms must be added in lexicographic ascending order.");
+        writeTermData(termRef);
+        writeTermToTrie(term);
+
+        maxLength = Math.max(maxLength, termRef.length);
+        swapTempWithPrevious();
+        pointId++;
+    }
+
+    private void writeTermToTrie(ByteComparable term) throws IOException
+    {
+        trieWriter.add(term, pointId);
+    }
+
+    private void writeTermData(BytesRef term) throws IOException
+    {
+        if ((pointId & TERMS_DICT_BLOCK_MASK) == 0)
+        {
+            offsetsWriter.add(termsOutput.getFilePointer() - bytesStartFP);
+
+            termsOutput.writeVInt(term.length);
+            termsOutput.writeBytes(term.bytes, term.offset, term.length);
+        }
+        else
+        {
+            final int prefixLength = StringHelper.bytesDifference(prevTerm.get(), term);
+            final int suffixLength = term.length - prefixLength;
+            assert suffixLength > 0: "terms must be unique";
+
+            termsOutput.writeByte((byte) (Math.min(prefixLength, 15) | (Math.min(15, suffixLength - 1) << 4)));
```

Comment:
nit: Here and in the equivalent section in the reader cursor, might be nice to have a quick explanation around packing the lengths into the first byte and the conditional vint...